File: app/src/main/java/com/example/viper/presentation/modules/rx2interactor/view/SampleRx2Fragment.java
package com.example.viper.presentation.modules.rx2interactor.view;

import android.databinding.DataBindingUtil;
import android.databinding.Observable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.viper.BR;
import com.example.viper.R;
import com.example.viper.application.AppDelegate;
import com.example.viper.data.SampleObject;
import com.example.viper.databinding.FragmentSampleRx2Binding;
import com.example.viper.presentation.common.dbadapter.ListConfig;
import com.example.viper.presentation.common.dbadapter.adapter.BindableAdapter;
import com.example.viper.presentation.common.dbadapter.listener.ActionClickListener;
import com.example.viper.presentation.modules.rx2interactor.assembly.SampleRx2Component;
import com.example.viper.presentation.modules.rx2interactor.assembly.SampleRx2DaggerModule;
import com.example.viper.presentation.modules.rx2interactor.contract.SampleRx2VmContract;
import com.example.viper.presentation.modules.rx2interactor.view.adapters.ActionType;
import com.example.viper.presentation.modules.rx2interactor.view.adapters.DataDelegate;
import com.vershininds.mixture.view.AbstractFragment;

import java.util.List;


public class SampleRx2Fragment extends AbstractFragment<SampleRx2VmContract.ViewModel, SampleRx2VmContract.Presenter> implements ActionClickListener {

    public static final String TAG = SampleRx2Fragment.class.getSimpleName();

    private SampleRx2Component diComponent;
    private FragmentSampleRx2Binding mBinding;

    private BindableAdapter<List<SampleObject>> mAdapter;

    public static Fragment newInstance() {
        return new SampleRx2Fragment();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_sample_rx2, container, false);

        mAdapter = new BindableAdapter<>(new DataDelegate(this));

        mBinding.setViewModel(getViewModel());

        ListConfig listConfig = new ListConfig.Builder(mAdapter)
                .setItemAnimator(new DefaultItemAnimator())
                .setHasFixedSize(true)
                .setDefaultDividerEnabled(true)
                .build(getActivity());

        mAdapter.setItems(getViewModel().getData());
        listConfig.applyConfig(getActivity(), mBinding.rvData);

        return mBinding.getRoot();
    }

    @Override
    protected void injectDi() {
        diComponent = AppDelegate.get()
                .presentationComponents()
                .sample2Component(new SampleRx2DaggerModule());
    }

    @Override
    protected SampleRx2VmContract.Presenter createPresenter() {
        return diComponent.getPresenter();
    }

    @Override
    protected void beforeViewAttached() {
        super.beforeViewAttached();
        getViewModel().addOnPropertyChangedCallback(vmObserver);
    }

    @Override
    protected void afterViewDetached() {
        super.afterViewDetached();
        getViewModel().removeOnPropertyChangedCallback(vmObserver);
    }


    private Observable.OnPropertyChangedCallback vmObserver = new Observable.OnPropertyChangedCallback() {
        @Override
        public void onPropertyChanged(Observable observable, int i) {
            SampleRx2VmContract.ViewModel vm = (SampleRx2VmContract.ViewModel) observable;
            if (i == BR.data) {
                mAdapter.setItems(vm.getData());
                mAdapter.notifyDataSetChanged();
            } else if (i == BR.error) {
                mBinding.txtError.setText(vm.getError());
            }
        }
    };

    @Override
    public void onActionClick(View view, String actionType, Object model) {
        switch (actionType) {
            case ActionType.CLICK:
                getPresenter().itemSelected((SampleObject)model);
                break;
            default:
                break;
        }
    }
}

File: app/src/main/java/com/vershininds/mixture/sample/presentation/modules/rx2interactor/view/SampleRx2Fragment.kt
package com.vershininds.mixture.sample.presentation.modules.rx2interactor.view

import android.arch.lifecycle.ViewModelProviders
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v7.widget.DefaultItemAnimator
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.vershininds.mixture.action.ViewAction
import com.vershininds.mixture.action.handle
import com.vershininds.mixture.action.subscribe
import com.vershininds.mixture.router.MviRouter
import com.vershininds.mixture.sample.R
import com.vershininds.mixture.sample.application.AppDelegate
import com.vershininds.mixture.sample.data.SampleObject
import com.vershininds.mixture.sample.databinding.FragmentSampleRx2Binding
import com.vershininds.mixture.sample.presentation.common.dbadapter.ListConfig
import com.vershininds.mixture.sample.presentation.common.dbadapter.adapter.BindableAdapter
import com.vershininds.mixture.sample.presentation.common.dbadapter.listener.ActionClickListener
import com.vershininds.mixture.sample.presentation.modules.rx2interactor.contract.SampleRx2RouterContract
import com.vershininds.mixture.sample.presentation.modules.rx2interactor.contract.SampleRx2VmContract
import com.vershininds.mixture.sample.presentation.modules.rx2interactor.di.SampleRx2Component
import com.vershininds.mixture.sample.presentation.modules.rx2interactor.di.SampleRx2DiModule
import com.vershininds.mixture.sample.presentation.modules.rx2interactor.view.adapters.ActionType
import com.vershininds.mixture.sample.presentation.modules.rx2interactor.view.adapters.DataDelegate
import com.vershininds.mixture.sample.presentation.modules.rx2interactor.viewmodel.SampleRx2Vm
import com.vershininds.mixture.sample.presentation.modules.rx2interactor.viewmodel.SampleRx2VmFactory
import com.vershininds.mixture.view.AndroidComponent
import com.vershininds.mixture.viewmodel.DataModel
import javax.inject.Inject

class SampleRx2Fragment : Fragment(), AndroidComponent, ActionClickListener {


    companion object {

        val TAG = SampleRx2Fragment::class.java.simpleName

        fun newInstance(): Fragment {
            return SampleRx2Fragment()
        }
    }

    private lateinit var diComponent: SampleRx2Component
    private lateinit var vm: SampleRx2Vm
    private lateinit var binding: FragmentSampleRx2Binding
    private lateinit var adapter: BindableAdapter<List<SampleObject>>

    @Inject
    internal lateinit var router: MviRouter<SampleRx2RouterContract.TypeRouterAction>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        diComponent = AppDelegate.get()
                .presentationComponents()
                .sample2Component(SampleRx2DiModule())
                .apply { inject(this@SampleRx2Fragment) }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
            inflater.inflate(R.layout.fragment_sample_rx2, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding = DataBindingUtil.bind<FragmentSampleRx2Binding>(view)!!

        adapter = BindableAdapter(DataDelegate(this))

        ListConfig.Builder(adapter)
                .setItemAnimator(DefaultItemAnimator())
                .setHasFixedSize(true)
                .setDefaultDividerEnabled(true)
                .build(activity)
                .applyConfig(activity, binding.rvData)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        val vmFactory = SampleRx2VmFactory(diComponent.getInteractor())

        vm = ViewModelProviders.of(this, vmFactory)
                .get(SampleRx2Vm::class.java)
                .apply { restoreInstanceState(savedInstanceState) }

        observeAction(vm)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(vm.saveInstanceState(outState))
    }

    override fun getSupportFragmentManager(): FragmentManager? {
        return fragmentManager
    }

    override fun onActionClick(view: View, actionType: String, model: Any) {
        when (actionType) {
            ActionType.CLICK -> vm.clickOnItem(model as SampleObject)
        }
    }

    private fun observeAction(model: SampleRx2Vm) {
        model.actionDispatcher.subscribe(viewLifecycleOwner) { action ->
            when (action) {
                is ViewAction -> handleViewAction(action)
                is SampleRx2RouterContract.TypeRouterAction -> handleRouterAction(action)
            }
        }
    }

    private fun handleViewAction(action: ViewAction) {
        action.handle {
            when (it) {
                is SampleRx2VmContract.TypeViewAction.DataAction -> handleDataAction(it)
/*                is SampleRx2VmContract.TypeViewAction.ErrorDialogWithCustomMessageAction ->
                    showErrorDialog(it.errorMsg)*/
            }
        }
    }

    private fun handleDataAction(action: SampleRx2VmContract.TypeViewAction.DataAction) {
        val dataModel = action.data

        val state = dataModel.state
        binding.state = state

        when (state) {
            DataModel.State.LOADING -> { /*nop*/
            }
            DataModel.State.ERROR -> {
                binding.txtError.text = dataModel.error
            }
            DataModel.State.DATA -> {
                adapter.items = dataModel.data
                adapter.notifyDataSetChanged()
            }
        }
    }

    private fun handleRouterAction(action: SampleRx2RouterContract.TypeRouterAction) {
        action.handle { router.actionHandler(this, it) }
    }
}